Add comprehensive type annotations

wch opened the pull request:
This PR follows up on a discussion I had with @mmarchetti. It does the following: 
- Turns on strict type checking with pyright.
- Adds many type annotations. These by and large should not affect run-time behavior, and so should be safe.

In the changes here, there are still many, many issues flagged by pyright in strict mode.
The purpose of this is to set things up for a second pull request, which will be made against this pull request instead of master. The second PR will have changes that are not quite as safe as this PR -- it will add more type annotations and makes some changes to the logic, and is therefore not as safe as this PR. The reason they're done with two separate pull requests is so that the substantive changes in the second PR are more visible and not drowned out by the changes in this PR, which are safe. This will make the second PR easier to review. 

After the second PR is merged into this one and tests are passing, we can merge this PR into master.

The overall goal is to get the code passing with pyright in strict mode in CI, and disable mypy. Once this is all done, it will be much easier to contribute to this repository and refactor the code here.

mmarchetti commented:
This is awesome! A huge step - no, actually a leap! - forward toward modernizing this codebase.

mmarchetti commented on rsconnect/actions.py:
```diff
@@ -488,767 +265,16 @@ def validate_quarto_engines(inspect):
-    return deploy_app(app_mode=AppModes.PYTHON_API, **locals())
-
-
-def deploy_python_fastapi(
```
This function is imported by `vetiver-python` and its removal may be why the tests are failing.

wch replied:
Ah, that's good to know! This makes me a little concerned, though: I removed many of the functions in `actions.py` because they were not used internally and seemed to essentially be duplicates of functions in `main.py`. Do we need to keep any other of those functions around in case someone external was using them?

Also, in this particular case, I wonder if we can just have `actions.deploy_python_fastapi` be an alias to `main.deploy_python_fastapi`.

mmarchetti replied:
Currently the only non-CLI clients are `vetiver-python` and `rsconnect-jupyter`.

https://github.com/rstudio/rsconnect-jupyter/blob/0c3d679d9ec799330134aa7710646870ec35e9c9/rsconnect_jupyter/__init__.py#L13C1-L29C45

mmarchetti replied:
> I wonder if we can just have `actions.deploy_python_fastapi` be an alias to `main.deploy_python_fastapi`

Yes, as long as we're careful to match the type signature that Vetiver is expecting.

wch replied:
Hm, there currently isn't a `main.deploy_python_fastapi`; the command line command is generated by the following and isn't saved in a variable anywhere:
```py
generate_deploy_python(app_mode=AppModes.PYTHON_FASTAPI, alias="fastapi", min_version="2021.08.0")
```

Also, the function signature is somewhat different and I don't want to risk breaking things, so for now I'll just bring back this code.

wch commented:
From some testing I did in #571, it looks like commit c476d69 made the test break. It looks like something on the server side stopped working? 

In that commit I added many type annotations, but didn't make any functional changes. I suppose it's possible that some of the syntax I used there was not compatible with Python 3.8.

This is the line that cause the error:

```
python vetiver-testing/setup-rsconnect/dump_api_keys.py vetiver-testing/rsconnect_api_keys.json
```

As far as I can tell, that code does not cause `rsconnect` to be imported at all, at least not on the client side. But the server side gives errors in response to requests. Does `rsconnect` get loaded on the server side somehow?

wch commented:
Now when I run the tests, the results are inconsistent. For example, on the same commit, I ran the workflow twice, and the failure happened differently between the two runs.

- First run: https://github.com/rstudio/rsconnect-python/actions/runs/8562649287/job/23466276020
- Second run: https://github.com/rstudio/rsconnect-python/actions/runs/8562649287/job/23466519638

I don't understand why the first one failed, but the second run failed in a way that makes sense to me and that I can fix.

wch commented:
I fixed the bug mentioned in the previous comment, and now here's a case where it fails on the first run and succeeds on the second:

1. https://github.com/rstudio/rsconnect-python/actions/runs/8562909967/job/23467049557
2. https://github.com/rstudio/rsconnect-python/actions/runs/8562909967/job/23467122442

I see that in #569, where the command was switched from `docker-compose` to `docker compose`, it had the same failure on the first attempt and then passed on the second attempt, so that makes me more confident that the failures aren't related to the content of this PR.

1. https://github.com/rstudio/rsconnect-python/actions/runs/8528435857/job/23362057518
2. https://github.com/rstudio/rsconnect-python/actions/runs/8528435857/job/23365251904


I suspect there's a timing problem with line `docker compose up -d` and the commands that come after it -- it might not have enough time to reliably start the container before the subsequent requests are made to the container. In 916535fd3433da8ce13993379b018072af78b31d I added a `sleep 4` after that line, and now it seems to be running more reliably -- two successes in a row.

aronatkins commented:
The timing problem is likely https://github.com/rstudio/rsconnect-python/issues/448. CC @tdstein
